include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,506 @@
+                    if (__start < __n && __local_id != 0 && __keys[__start] != __keys[__start - 1])
+                        __max_end = __local_id;
+
+                    for (int32_t __i = __start; __i < __end; ++__i)
+                    {
+                        if (__first)
+                        {
+                            __accumulator = __values[__i];
+                            __first = false;
+                        }
+                        else
+                            __accumulator = __binary_op(__accumulator, __values[__i]);
+
+                        __out_values[__i] = __accumulator;
+
+                        // clear the accumulator if we reach end of segment
+                        if (__n - 1 == __i || __keys[__i] != __keys[__i + 1])
+                        {
+                            __accumulator = {};
+                            __max_end = __local_id;
+                            __first = true;
+                        }
+                    }
+
+                    __loc_acc[__local_id] = __accumulator;
+
+                    // 1b. Perform a work group scan to find the carry in value to apply to each item.
+                    auto __closest_seg_id = sycl::inclusive_scan_over_group(__group, __max_end, sycl::maximum<>());
+
+                    __val_type __carry_in =
+                        wg_segmented_scan(__item, __loc_acc, __local_id, __local_id - __closest_seg_id, __accumulator,
+                                          __binary_op, __wgroup_size);
+
+                    // 1c. Update partial scan values and write to global memory.
+                    if (__local_id != 0 && __start < __n && __keys[__start] == __keys[__start - 1])
+                    {
+                        for (int32_t __i = __start; __i < __end; ++__i)
+                        {
+                            __out_values[__i] = __binary_op(__carry_in, __out_values[__i]);
+
+                            if (__i == __n - 1 || __keys[__i] != __keys[__i + 1])
+                                break;
+                        }
+                    }
+                    
+                    // 1d. Store partials and presence of segment flags for use in phase 2
+                    if (__local_id == __wgroup_size - 1) // last work item writes the group's carry out
+                    {
+                        auto __group_id = __group.get_group_id(0);
+
+                        // If no segment ends in the item, the aggregates from previous work groups must be applied.
+                        if (__max_end == 0)
+                            __partials_acc[__group_id] =
+                                __binary_op(__carry_in, __accumulator); // inclusive with last element
+
+                        else
+                            __partials_acc[__group_id] = __accumulator;
+
+                        bool __seg_present = static_cast<bool>(__closest_seg_id);
```

Comment:
can we avoid `static_cast` by writing `bool __seg_present = __closest_seg_id != 0;` ?

Reply:
Yes, I will make this change